cudax/include/cuda/experimental/__container/uninitialized_async_buffer.cuh

```diff
@@ -93,7 +96,28 @@ private:
     constexpr size_t __alignment = alignof(_Tp);
     size_t __space               = __get_allocation_size(__count_);
     void* __ptr                  = __buf_;
-    return reinterpret_cast<_Tp*>(_CUDA_VSTD::align(__alignment, __count_ * sizeof(_Tp), __ptr, __space));
+    return _CUDA_VSTD::launder(
+      reinterpret_cast<_Tp*>(_CUDA_VSTD::align(__alignment, __count_ * sizeof(_Tp), __ptr, __space)));
+  }
+
+  //! @brief Causes the buffer to be treated as a span when passed to cudax::launch.
+  //! @pre The buffer must have the cuda::mr::device_accessible property.
+  _CCCL_NODISCARD_FRIEND _CUDA_VSTD::span<_Tp>
+  __cudax_launch_transform(::cuda::stream_ref, uninitialized_async_buffer& __self) noexcept
```

Comment:
I am unsure, in case the streams are different do we want to synchronize here or in a central place?

Reply:
Seems like this could lead to unnecessary extra synchronization. You don't know if the stream the buffer was last allocated/written on needs to be synchronized. It may have been already.

E.g.
 
```
auto buf = buffer(size, stream_a);
launch(kernel, stream_a, buffer); // initialize buffer with computation in kernel (no sync)
stream_a.wait(); // sync stream_a

// Launch 4 instances of kernel to operate on 4 different buffers on 4 streams. 
// All kernels read `buf` as an input. 
// The suggested sync in `__cudax_launch_transform()` would synchronize all 4 streams before launching
// no streams need to be synced in this loop
for (int i = 0; i < 4; i++) {
  launch(kernel, streams[i], buffers[i], buf)
}
```

Reply:
Yeahy, but that is the same discussion as about lifetimes. We dont know whether a resource might go out of scope, so we need to do the safe thing

Reply:
Do you have a good example where the user of the buffer is unable to synchronize themselves? If one chooses to use an `async` buffer, they should be aware that they may need to do some synchronization. If we assume that the user doesn't know what they are doing then we don't give them the ability to hit SOL.

Reply:
I think the protocol for `__cudax_launch_transform` is to return a wrapper here that can covert to a span. Then synchronize here and synchronize back in the destructor of that wrapper.

We also definitely need an opt-out of the synchronization, but not sure how it would look like. Something like `cudax::skip_sync(buffer)`, we should try to come up with something generic for other similar cases.